src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -0,0 +1,1241 @@
+        /// </remarks>
+        /// <exception cref="LockRecursionException">
+        /// The lock has reached the limit of recursive enters. The limit is implementation-defined, but is expected to be high
+        /// enough that it would typically not be reached when the lock is used properly.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public void Enter()
+        {
+            ThreadId currentThreadId = TryEnter_Inlined(timeoutMs: -1);
+            Debug.Assert(currentThreadId.IsInitialized);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private ThreadId EnterAndGetCurrentThreadId()
+        {
+            ThreadId currentThreadId = TryEnter_Inlined(timeoutMs: -1);
+            Debug.Assert(currentThreadId.IsInitialized);
+            Debug.Assert(currentThreadId.Id == _owningThreadId);
+            return currentThreadId;
+        }
+
+        /// <summary>
+        /// Enters the lock and returns a <see cref="Scope"/> that may be disposed to exit the lock. Once the method returns,
+        /// the calling thread would be the only thread that holds the lock. This method is intended to be used along with a
+        /// language construct that would automatically dispose the <see cref="Scope"/>, such as with the C# <code>using</code>
+        /// statement.
+        /// </summary>
+        /// <returns>
+        /// A <see cref="Scope"/> that may be disposed to exit the lock.
+        /// </returns>
+        /// <remarks>
+        /// If the lock cannot be entered immediately, the calling thread waits for the lock to be exited. If the lock is
+        /// already held by the calling thread, the lock is entered again. The calling thread should exit the lock, such as by
+        /// disposing the returned <see cref="Scope"/>, as many times as it had entered the lock to fully exit the lock and
+        /// allow other threads to enter the lock.
+        /// </remarks>
+        /// <exception cref="LockRecursionException">
+        /// The lock has reached the limit of recursive enters. The limit is implementation-defined, but is expected to be high
+        /// enough that it would typically not be reached when the lock is used properly.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Scope EnterScope() => new Scope(this, EnterAndGetCurrentThreadId());
+
+        /// <summary>
+        /// A disposable structure that is returned by <see cref="EnterScope()"/>, which when disposed, exits the lock.
+        /// </summary>
+        public ref struct Scope
+        {
+            private Lock? _lockObj;
+            private ThreadId _currentThreadId;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal Scope(Lock lockObj, ThreadId currentThreadId)
+            {
+                _lockObj = lockObj;
+                _currentThreadId = currentThreadId;
```

Comment:
Interesting optimization. `ref` structs cannot be on the GC heap, but it is not strictly guaranteed that they are thread-static. That may hold now, but may not hold in the future. 
For example there is a lot of desire and some proposals that could allow ref structs in async methods. If/when that is allowed, ref structs will stay within the same logical control flow but the thread identity may change around `await` and that could break this optimization.

I tried to measure the impact of this optimization and it is measurable, although relatively small (about 5% on targeted microbenchmark). I think that is because when Enter/Dispose are in the same method, JIT would hoist the access to the threadstatic, so the second lookup might not be as expensive as the first.

It is hard to make a call on this, considering there seems to be an improvement, but the optimization is technically unsafe.  
Also it must be considered that using locks in async code is by itself is a fragile practice.

Reply:
> I tried to measure the impact of this optimization and it is measurable, although relatively small (about 5% on targeted microbenchmark). I think that is because when Enter/Dispose are in the same method, JIT would hoist the access to the threadstatic, so the second lookup might not be as expensive as the first.

The relevant enter/exit methods are no-inlined, so that wouldn't be it. What I found is that some processors are pretty good at prefetching through the second thread-static lookup when not much is done inside the lock. It doesn't take much though to break the prefetching though, even a small amount of code in the lock like manipulating a simple collection leads to improvements around 5-10%.

> For example there is a lot of desire and some proposals that could allow ref structs in async methods. If/when that is allowed, ref structs will stay within the same logical control flow but the thread identity may change around `await` and that could break this optimization.

I can see ref structs being allowed in async methods. Whether it would be allowed around awaits though I'm not sure. If there's a desire to allow them around awaits I imagine that would be somewhat of a breaking change anyway, as existing ref structs that rely on current semantics can break. We can undo the optimization then, or hopefully there would be an alternate way to retain the current semantics, maybe there could be a new syntax for ref structs that can be used around awaits or something else.